source/common/common/shared_token_bucket_impl.cc

```diff
@@ -0,0 +1,49 @@
+#include "common/common/shared_token_bucket_impl.h"
+
+#include <chrono>
+
+#include "common/common/lock_guard.h"
+
+namespace Envoy {
+
+const char SharedTokenBucketImpl::GetImplSyncPoint[] = "pre_get_impl";
+const char SharedTokenBucketImpl::ResetCheckSyncPoint[] = "post_reset_check";
+
+SharedTokenBucketImpl::SharedTokenBucketImpl(uint64_t max_tokens, TimeSource& time_source,
+                                             double fill_rate)
+    : impl_(max_tokens, time_source, fill_rate), reset_once_(false) {}
+
+uint64_t SharedTokenBucketImpl::consume(uint64_t tokens, bool allow_partial)
+/*ABSL_EXCLUSIVE_LOCKS_REQUIRED(mutex_)*/ {
+  Thread::LockGuard lock(mutex_);
+  synchronizer_.syncPoint(GetImplSyncPoint);
+  return getImpl().consume(tokens, allow_partial);
+};
+
+std::chrono::milliseconds SharedTokenBucketImpl::nextTokenAvailable()
+/*ABSL_EXCLUSIVE_LOCKS_REQUIRED(mutex_)*/ {
+  Thread::LockGuard lock(mutex_);
+  synchronizer_.syncPoint(GetImplSyncPoint);
+  return getImpl().nextTokenAvailable();
+};
+
+void SharedTokenBucketImpl::reset(uint64_t num_tokens) /*ABSL_EXCLUSIVE_LOCKS_REQUIRED(mutex_)*/ {
+  Thread::LockGuard lock(mutex_);
+  // Don't reset if reset once before.
+  if (reset_once_) {
```

Comment:
Why is this behavior change from `TokenBucketImpl` needed?

Reply:
The reset method is used by [streamratelimiter ](https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/extensions/filters/http/fault/fault_filter.cc#L542) to achieve smooth token flow at the start of a stream. When the tokenbucket is not shared and created per stream, it works fine. But since this bucket is shared across streams, we only need the reset the very first time to achieve the smooth token flow for this bucket. I prefer to have it here since the caller code is oblivious to whether the bucket is shared.

Reply:
Since I'm already touching the interface, I guess I can add a new interface method or a new parameter to this method to make it explicit.